### PR TITLE
refactor(tooltip): Extract the "one tooltip at a time" logic

### PR DESCRIPTION
**Pull request by shane-tomlinson**
This logic is used in #6273 to ensure that no tooltips are displayed
whenever the password strength bubble is displayed.

@vbudhram - r?

### CODE REVIEW

File: app/scripts/views/mixins/one-of-type-mixin.js [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+      if (entry && entry.view === this) {
+        visibleViews[type] = null;
+        delete visibleViews[type];
+      }
+    },
+
+    // wrap the triggering function so that
+    // housekeeping is automatic.
+    [trigger] () {
+      const entry = visibleViews[type];
```
Comment (vbudhram) [inline app/scripts/views/mixins/one-of-type-mixin.js]: Is it ok to fall through here?

File: app/scripts/views/mixins/one-of-type-mixin.js [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+
+/**
+ * Create the mixin.
+ *
+ * @export
+ * @param {Object} options
+ * @param {String|Function} options.action - action to call to hide the view.
+ *   If a string, this[action]() will be called, if a function, the function
+ *   will be called with the current view as the context.
+ * @param {String} options.trigger - name of the method that
```
Comment (vbudhram) [inline app/scripts/views/mixins/one-of-type-mixin.js]: from pair review: This should be causes hidden views to be visible.

File: app/scripts/views/mixins/one-of-type-mixin.js [hunk 3]
```diff
@@ -0,0 +1,57 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+/**
+ * A mixin that only allows one of the given type to be visible
+ * at a time. Useful for things like tooltips where only
+ * one should be visible at a time.
```
Comment (shane-tomlinson) [inline app/scripts/views/mixins/one-of-type-mixin.js]: Expand on this comment about how this is used w/ the password strength balloon

File: app/scripts/views/mixins/one-of-type-mixin.js [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+ *   causes visible views to be hidden.
+ * @param {String} options.type - namespace. Only one item per namespace
+ *   can be visible at a time.
+ * @returns {Function}
+ */
+export default function (options) {
+  const { action, trigger, type } = options;
+  return {
+    beforeDestroy () {
+      // The current view can no longer be active after being destroyed
```
Comment (shane-tomlinson) [inline app/scripts/views/mixins/one-of-type-mixin.js]: `active` => `visible`

File: app/scripts/views/mixins/one-of-type-mixin.js [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+      const entry = visibleViews[type];
```
Comment (vbudhram) [inline app/scripts/views/mixins/one-of-type-mixin.js]: Could this called `view`?

File: app/scripts/views/tooltip.js [hunk 6]
```diff
@@ -9,10 +9,11 @@
 import $ from 'jquery';
 import _ from 'underscore';
 import BaseView from './base';
+import Cocktail from 'cocktail';
 import KeyCodes from '../lib/key-codes';
+import OneOfTypeMixin from './mixins/one-of-type-mixin';
```
Comment (vbudhram) [inline app/scripts/views/tooltip.js]: Would `one-visible-view-per-type` be a little clear?

File: app/scripts/views/mixins/one-of-type-mixin.js [hunk 7]
```diff
@@ -0,0 +1,52 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+/**
+ * A mixin that only allows one of the given type to be visible
```
Comment (vbudhram) [inline app/scripts/views/mixins/one-of-type-mixin.js]: `one of the given type` -> `only one view of a certain type`?

**Pull request merged**